src/platform/qt/CoreController.cpp

```diff
@@ -0,0 +1,725 @@
+
+		controller->m_activeBuffer->fill(0xFF);
+		controller->finishFrame();
+	};
+
+	m_threadContext.frameCallback = [](mCoreThread* context) {
+		CoreController* controller = static_cast<CoreController*>(context->userData);
+
+		controller->finishFrame();
+	};
+
+	m_threadContext.cleanCallback = [](mCoreThread* context) {
+		CoreController* controller = static_cast<CoreController*>(context->userData);
+
+		controller->clearMultiplayerController();
+		QMetaObject::invokeMethod(controller, "stopping");
+	};
+
+	m_threadContext.logger.d.log = [](mLogger* logger, int category, enum mLogLevel level, const char* format, va_list args) {
+		mThreadLogger* logContext = reinterpret_cast<mThreadLogger*>(logger);
+		mCoreThread* context = logContext->p;
+
+		static const char* savestateMessage = "State %i loaded";
+		static const char* savestateFailedMessage = "State %i failed to load";
+		static int biosCat = -1;
+		static int statusCat = -1;
+		if (!context) {
+			return;
+		}
+		CoreController* controller = static_cast<CoreController*>(context->userData);
+		QString message;
+		if (biosCat < 0) {
+			biosCat = mLogCategoryById("gba.bios");
+		}
+		if (statusCat < 0) {
+			statusCat = mLogCategoryById("core.status");
+		}
+#ifdef M_CORE_GBA
+		if (level == mLOG_STUB && category == biosCat) {
+			va_list argc;
+			va_copy(argc, args);
+			int immediate = va_arg(argc, int);
+			va_end(argc);
+			QMetaObject::invokeMethod(controller, "unimplementedBiosCall", Q_ARG(int, immediate));
+		} else
+#endif
+		if (category == statusCat) {
+			// Slot 0 is reserved for suspend points
+			if (strncmp(savestateMessage, format, strlen(savestateMessage)) == 0) {
+				va_list argc;
+				va_copy(argc, args);
+				int slot = va_arg(argc, int);
+				va_end(argc);
+				if (slot == 0) {
+					format = "Loaded suspend state";
+				}
+			} else if (strncmp(savestateFailedMessage, format, strlen(savestateFailedMessage)) == 0) {
+				va_list argc;
+				va_copy(argc, args);
+				int slot = va_arg(argc, int);
+				va_end(argc);
+				if (slot == 0) {
+					return;
+				}
+			}
+			message = QString().vsprintf(format, args);
+			QMetaObject::invokeMethod(controller, "statusPosted", Q_ARG(const QString&, message));
+		}
+		if (level == mLOG_FATAL) {
+			mCoreThreadMarkCrashed(controller->thread());
+			QMetaObject::invokeMethod(controller, "crashed", Q_ARG(const QString&, QString().vsprintf(format, args)));
+		}
+		message = QString().vsprintf(format, args);
+		QMetaObject::invokeMethod(controller, "logPosted", Q_ARG(int, level), Q_ARG(int, category), Q_ARG(const QString&, message));
+	};
+}
+
+CoreController::~CoreController() {
+	endVideoLog();
+	stop();
+	disconnect();
+
+	if (m_tileCache) {
+		mTileCacheDeinit(m_tileCache.get());
+		m_tileCache.reset();
+	}
+
+	mCoreThreadJoin(&m_threadContext);
+
+	mCoreConfigDeinit(&m_threadContext.core->config);
+	m_threadContext.core->deinit(m_threadContext.core);
+}
+
+color_t* CoreController::drawContext() {
+	QMutexLocker locker(&m_mutex);
```

Comment:
Yeah, so what's going on here is that `m_completeBuffer` is a rudimentary double-buffer--it's swapped between two objects in a different thread so here we just want one of them. The `data()` should remain valid since they are never resized after the initial sizing (in the constructor). Although that may change in the future (e.g. when I add SGB support, hm). That said, there are only two threads it'll ever be accessed in, so I think that's overkill.